src/lib/shim/src/lib.rs

```diff
@@ -24,9 +34,191 @@ pub fn manager_shmem() -> Option<&'static ManagerShmem> {
+            unsafe {
+                rustix::mm::munmap(
+                    stack_to_free_now as *mut core::ffi::c_void,
+                    SHIM_SIGNAL_STACK_SIZE,
+                )
+            }
+            .unwrap();
+        }
+    }
+}
+
+mod tls_ipc {
+    use super::*;
+    static IPC_DATA_BLOCK: ShimTlsVar<RefCell<Option<ShMemBlockAlias<IPCData>>>> =
+        ShimTlsVar::new(|| RefCell::new(None));
+
+    // Panics if this thread's IPC hasn't been initialized yet.
+    pub fn with<O>(f: impl FnOnce(&IPCData) -> O) -> O {
+        let ipc = IPC_DATA_BLOCK.get();
+        let ipc = ipc.borrow();
+        ipc.as_ref().map(|block| f(block)).unwrap()
+    }
+
+    /// # Safety
+    ///
+    /// `blk` must contained a serialized block of
+    /// type `IPCData`, which outlives the current thread.
+    pub unsafe fn set(blk: &ShMemBlockSerialized) {
+        let blk: ShMemBlockAlias<IPCData> = unsafe { Serializer::global().deserialize(blk) };
+        assert!(IPC_DATA_BLOCK.get().replace(Some(blk)).is_none());
+    }
+
+    /// # Safety
+    ///
+    /// Environment variable SHADOW_IPC_BLK must contained a serialized block of
+    /// type `IPCData`, which outlives the current thread.
+    pub unsafe fn set_from_env() {
+        use core::fmt::Write;
+        let mut envname = FormatBuffer::<20>::new();
+        write!(&mut envname, "SHADOW_IPC_BLK").unwrap();
+        // SAFETY: Kind of not. We should pass this some other way than an environment
+        // variable. Ok in practice as long as we do our initialization after libc's early
+        // initialization, and nothing else mutates this environment variable.
+        // https://github.com/shadow/shadow/issues/2848
+        let ipc_blk = unsafe { libc::getenv(envname.as_cstr().unwrap().as_ptr()) };
+        assert!(!ipc_blk.is_null());
+        let ipc_blk = unsafe { CStr::from_ptr(ipc_blk) };
+        let ipc_blk = core::str::from_utf8(ipc_blk.to_bytes()).unwrap();
+
+        let ipc_blk = ShMemBlockSerialized::decode_from_string(ipc_blk).unwrap();
+        set(&ipc_blk);
+    }
+}
+
+mod tls_thread_shmem {
+    use super::*;
+
+    static THREAD_SHMEM: ShimTlsVar<RefCell<Option<ShMemBlockAlias<ThreadShmem>>>> =
+        ShimTlsVar::new(|| RefCell::new(None));
+
+    // Panics if not initialized yet.
+    pub fn with<O>(f: impl FnOnce(&ThreadShmem) -> O) -> O {
+        let thread = THREAD_SHMEM.get();
+        let thread = thread.borrow();
+        thread.as_ref().map(|block| f(block)).unwrap()
+    }
+
+    /// # Safety
+    ///
+    /// `blk` must contained a serialized block of
+    /// type `ThreadShmem`, which outlives the current thread.
+    pub unsafe fn set(blk: &ShMemBlockSerialized) {
+        let blk: ShMemBlockAlias<ThreadShmem> = unsafe { Serializer::global().deserialize(blk) };
+        assert!(THREAD_SHMEM.get().replace(Some(blk)).is_none());
+    }
+}
+
 mod shimlogger;
 pub use shimlogger::export as shimlogger_export;
 
+mod tls;
```

Comment:
optional nit: I think it's nicer when the `mod` statements that link to other files are at the top of the file so they're all in one place, and it's easy to add more.